internal/pkg/agent/application/actions/handlers/handler_helpers.go

```diff
@@ -0,0 +1,166 @@
+type backoffActionDispatcher struct {
+	log           *logp.Logger
+	performAction performActionFunc
+
+	timeout    time.Duration
+	minBackoff time.Duration
+	maxBackoff time.Duration
+}
+
+const (
+	defaultActionDispatcherTimeout    = 40 * time.Second
+	defaultActionDispatcherBackoffMin = 500 * time.Millisecond
+	defaultActionDispatcherBackoffMax = 10 * time.Second
+)
+
+func newBackoffActionDispatcher(log *logp.Logger, performAction performActionFunc) backoffActionDispatcher {
+	return backoffActionDispatcher{
+		log:           log,
+		performAction: performAction,
+		timeout:       defaultActionDispatcherTimeout,
+		minBackoff:    defaultActionDispatcherBackoffMin,
+		maxBackoff:    defaultActionDispatcherBackoffMax,
+	}
+}
+
+func (d backoffActionDispatcher) Dispatch(ctx context.Context, action dispatchableAction, ucs []unitWithComponent) error {
+	if action == nil {
+		return nil
+	}
+
+	// Deserialize the action into map[string]interface{} for dispatching over to the apps
+	params, err := action.MarshalMap()
+	if err != nil {
+		return err
+	}
+
+	actionType := action.Type()
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	dispatch := func(uc unitWithComponent) error {
+		if uc.unit.Config == nil {
+			return nil
+		}
+		d.log.Debugf("Dispatch %v action to %v", actionType, uc.unit.Config.Type)
+		res, err := d.performAction(ctx, uc.component, uc.unit, uc.unit.Config.Type, params)
+		if err != nil {
+			d.log.Debugf("%v failed to dispatch to %v, err: %v", actionType, uc.component.ID, err)
+			return err
+		}
+
+		strErr := readMapString(res, "error", "")
+		if strErr != "" {
+			d.log.Debugf("%v failed for %v, err: %v", actionType, uc.component.ID, strErr)
+			return errors.New(strErr)
+		}
+		return nil
+	}
+
+	dispatchWithBackoff := func(uc unitWithComponent) error {
+		ctx, cn := context.WithTimeout(ctx, d.timeout)
+		defer cn()
+
+		attempt := 1
+		backExp := backoff.NewExpBackoff(ctx.Done(), d.minBackoff, d.maxBackoff)
+		start := time.Now()
+
+		for {
+			err := dispatch(uc)
+			if err != nil {
+				if backExp.Wait() {
+					d.log.Debugf("%v action dispatch to %v with backoff attempt: %v, after %v since start", actionType, uc.component.ID, attempt, time.Since(start))
+					attempt++
+					continue
+				}
+				return err
+			}
+			return nil
+		}
+	}
+
+	// Iterate through the components and dispatch the action is the action type is listed in the proxied_actions
+	for _, uc := range ucs {
+		// Send the action to the target unit via g.Go to collect any resulting errors
+		target := uc
+		g.Go(func() error {
+			return dispatchWithBackoff(target)
+		})
+	}
+
+	return g.Wait()
+}
+
+// dispatchActionInParallel dispatches actions to the units/components in parallel, with exponential backoff and timeout
+func dispatchActionInParallel(ctx context.Context, log *logp.Logger, action dispatchableAction, ucs []unitWithComponent, performAction performActionFunc) error {
+	return newBackoffActionDispatcher(log, performAction).Dispatch(ctx, action, ucs)
```

Comment:
What's the reasoning for making this a standalone object with a single method instead of just implementing the contents of `Dispatch` here? It looks like this object is never shared outside the package, so the public `Dispatch` is unneeded, and I worry this will be confused with the main `ActionDispatcher` which is a persistent global object that invokes all standard action handlers, where this is a one-off ephemeral struct that only exists for the duration of its call. Could we remove the struct wrapper and call this function e.g. `notifyUnitsOfProxiedAction`?

Reply:
wanted to keep the struct to allow parameterize the backoff/timeout, especially for testing, using shorter time durations. can rename.